staging/src/k8s.io/component-base/logs/options.go

```diff
@@ -74,14 +78,17 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 }
 
 // apply set klog logger from LogFormat type
-func (o *Options) apply() {
+func (o *Options) apply(featureGate featuregate.FeatureGate) {
 	// if log format not exists, use nil loggr
 	factory, _ := registry.LogRegistry.Get(o.Config.Format)
 	if factory == nil {
 		klog.ClearLogger()
 	} else {
-		log, flush := factory.Create(o.Config.Options)
-		klog.SetLoggerWithOptions(log, klog.FlushLogger(flush))
+		// This logger will do its own verbosity checking, using the exact same
+		// configuration as klog itself.
+		log, flush := factory.Create(o.Config)
+		// Therefore it can get called directly.
+		klog.SetLoggerWithOptions(log, klog.ContextualLogger(true), klog.FlushLogger(flush))
```

Comment:
Why pass `klog.ContextualLogger(true)` even though it will be override below with `klog.EnableContextualLogging(contextualLoggingEnabled)`?

Reply:
That depends on the intended purpose of the feature gate: we can either interpret it as "opt into potentially more expensive operations" (current code and the original concern that was raised during KEP review) or more generally as "don't do anything related to contextual logging". In the latter case we also need to pass false here to disable the JSON logger being called directly.

Let's be strict about it and also disable direct calls.

Reply:
I'm not sure I understand. Correct me if I'm wrong, contextual logging is a internal bool variable that can be switched in two ways: 1. by passing `klog.ContextualLogger(true)` to `klog.SetLoggerWithOptions` and 2. by calling `klog.EnableContextualLogging(contextualLoggingEnabled)`. 

I understand why we introduced two ways of configuring this in klog, however why we are using two different ways to configure it here? Would setting it once be more clearer?

Reply:
> contextual logging is a internal bool variable that can be switched in two ways

There are two different aspects:
- the current logger *can* get called directly (depends on verbosity check in logger)
- the current logger *may* get called directly and (feature gate)

That's why we have two different klog calls. In Kubernetes, the JSON logger support verbosity checks, so only the feature gate matters. Outside of Kubernetes, someone might use `EnableContextualLogging(true)` in combination with `ContextualLogger(false)`.